src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,573 @@
+	  head.last_committed_record_base = 0;
+	  head.written_to = head.start_offset;
+	  head.applied_to = 0;
+	  head.flag = 0;
+	  head.csum_type = 0;
+	  head.csum = 0;
+	  head.cur_segment_seq = 0;
+	  head.start = start_addr;
+	  head.end = end_addr;
+	  head.device_id = config.device_id;
+	  encode(head, bl);
+	} catch (ceph::buffer::error &e) {
+	  logger().debug("unable to encode super block from underlying deivce");
+	  return mkfs_ertr::now();
+	}
+	header = head;
+	logger().debug(
+	  "initialize header block in CBJournal, length {}",
+	  bl.length());
+	return device_write_bl(start_addr, bl
+	).handle_error(
+	  mkfs_ertr::pass_further{},
+	  crimson::ct_error::assert_all{
+	  "Invalid error device_write during CBJournal::mkfs"
+	});
+      }),
+      mkfs_ertr::pass_further{},
+      crimson::ct_error::assert_all{
+        "Invalid error read_super in CBJournal::mkfs"
+      }
+    );
+  }).handle_error(
+    mkfs_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+    "Invalid error _open_device in CBJournal::mkfs"
+  }).finally([this] {
+    if (device) {
+      return device->close();
+    } else {
+      return seastar::now();
+    }
+  });
+}
+
+CBJournal::open_for_write_ertr::future<> CBJournal::_open_device(
+        const std::string path)
+{
+  ceph_assert(device);
+  return device->open(path, seastar::open_flags::rw
+  ).handle_error(
+    open_for_write_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error device->open"
+    }
+  );
+}
+
+ceph::bufferlist CBJournal::encode_super()
+{
+  bufferlist bl;
+  encode(header, bl);
+  return bl;
+}
+
+CBJournal::open_for_write_ret CBJournal::open_for_write()
+{
+  return open_for_write(CBJOURNAL_START_ADDRESS);
+}
+
+CBJournal::close_ertr::future<> CBJournal::close()
+{
+  return sync_super(
+  ).safe_then([this]() -> close_ertr::future<> {
+    return device->close();
+  }).handle_error(
+    open_for_write_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error sync_super"
+    }
+  );
+}
+
+CBJournal::open_for_write_ret CBJournal::open_for_write(rbm_abs_addr start)
+{
+  return _open_device(path
+  ).safe_then([this, start]() {
+    return read_super(start
+    ).handle_error(
+      open_for_write_ertr::pass_further{},
+      crimson::ct_error::assert_all{
+	"Invalid error read_super"
+    }).safe_then([this](auto p) mutable {
+      auto &[head, bl] = *p;
+      header = head;
+      logger().debug(" super : {} ", header);
+      paddr_t paddr = convert_abs_addr_to_paddr(
+	get_written_to(),
+	header.device_id);
+      return open_for_write_ret(
+	open_for_write_ertr::ready_future_marker{},
+	journal_seq_t{
+	  header.cur_segment_seq,
+	  paddr
+	});
+    });
+  }).handle_error(
+    open_for_write_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error _open_device"
+  });
+}
+
+CBJournal::write_ertr::future<> CBJournal::append_record(
+  ceph::bufferlist bl,
+  rbm_abs_addr addr)
+{
```

Comment:
This method sequentially writes out the two portions of the record.  Is there any reason not to submit the two writes concurrently?